lib/Sema/CodeSynthesis.cpp

```diff
@@ -2405,9 +2405,16 @@ static void configureDesignatedInitAttributes(TypeChecker &tc,
 
   // If the superclass has its own availability, make sure the synthesized
   // constructor is only as available as its superclass's constructor.
-  if (superclassCtor->getAttrs().hasAttribute<AvailableAttr>()) {
+  {
+    SmallVector<Decl *, 2> asAvailableAs;
+    asAvailableAs.push_back(superclassCtor);
+    Decl *parentDecl = classDecl;
+    while (parentDecl != nullptr) {
+      asAvailableAs.push_back(parentDecl);
+      parentDecl = parentDecl->getDeclContext()->getAsDeclOrDeclExtensionContext();
```

Comment:
Looks like applyInferredAvailableAttrs() is only called twice, once here and again when propagating availability to accessors. I don't think the latter needs to do this, since storage being more available than its surrounding context is an error.

The only reason we have to applyInferredAvailableAttrs() with constructors is because the superclass availability plays a role. So I think we should not need to change applyInferredAvailableAttrs().

Reply:
Hm, I bet we're getting away with not looking at the superclass constructor's enclosing contexts because DIs are always in the main class body, and the main class body's availability is always broader than the subclass's availability. Can you add a comment to that effect and we'll call it good?